[now-build-utils] Add `handle: miss` route to zero config

styfle opened the pull request:
This PR adds the route `handle: miss` and a catch-all route for the api directory.

The plan is to rename files in `/api` and `/public` (in a future PR) to be extension-less and then the route is only used to rewrite the extension to the extension-less file (for example, `/api/user.go` => `/api/user`)

This reduces the routes needed for zero config (previously N routes for N files down to 1 route for N files).

AndyBitz commented on packages/now-build-utils/src/detect-routes.ts:
```diff
@@ -292,11 +292,16 @@ export async function detectRoutes(
     const directory = publicBuilder.src.replace('/**/*', '');
 
-    routesResult.defaultRoutes.push({
+    defaultRoutes.push({
```
I'd prefer `routesResult.defaultRoutes` since updating a deconstructed reference wouldn't be as clear when the code gets more complex

dav-is reviewed (changes requested):
This can't ship until now-proxy does, we also need to discuss API files being 1:1 with routes

styfle commented:
@dav-is I guess I need `check: true` on each of the zero config routes?

This is what was generated with this PR:

```json
"routes": [
    {
      "handle": "miss"
    },
    {
      "src": "^/api/([^\\/]+)/([^\\/]+)$",
      "dest": "/api/[endpoint]/[id].js?endpoint=$1&id=$2"
    },
    {
      "src": "^/api/([^\\/]+)$",
      "dest": "/api/[endpoint].js?endpoint=$1"
    },
    {
      "status": 404,
      "src": "/api(\\/.*)?$"
    }
  ],
```

dav-is commented:
@styfle yes, except the last one doesn't have a dest so it gets continue

styfle commented:
@dav-is I added `check: true` and `continue: true`.

I also changed some of the types so that `@now/routing-utils` is the correct source of route types instead of `@now/build-utils`.

Lastly, I updated the build script order so that `@now/routing-utils` is hoisted to the top.

Is `now-proxy` released yet?

styfle commented on packages/now-build-utils/src/detect-routes.ts:
```diff
@@ -83,7 +83,7 @@ function createRouteFromPath(filePath: string): Route {
   const dest = `/${filePath}${query.length ? '?' : ''}${query.join('&')}`;
 
-  return { src, dest };
+  return { src, dest, check: true };
```
@dav-is  Should the the `enableExperimentalHandleMiss` flag also affect the `check: true` property?

styfle committed:
[now-build-utils] Add `handle: miss` route to zero config

styfle committed:
Fix build

dav-is commented on packages/now-build-utils/src/detect-routes.ts:
```diff
@@ -265,11 +266,27 @@ async function detectApiRoutes(
+        },
+        {
+          status: 404,
+          src: '/api(\\/.*)?$',
```
We don't need to escape the `/` here or below

styfle replied:
Thanks, I removed the superfluous escape sequence

dav-is commented on packages/now-build-utils/src/detect-routes.ts:
```diff
@@ -289,18 +306,23 @@ function getPublicBuilder(builders: Builder[]): Builder | null {
+    if (featHandleMiss) {
+      route.check = true;
+    }
+    routesResult.defaultRoutes.push(route);
```
To confirm, this comes after handle: miss right?

styfle replied:
Indeed, but I updated the code to make it more clear

dav-is commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -570,7 +570,8 @@ export default class DevServer {
           files,
-          builders
+          builders,
+          false // TODO: enable after implementing `handle: miss`
```
We need to tie this to a temporary now.json option

styfle replied:
I am not implementing `now dev` yet so I'll set this to default as `false` and avoid changing this file.

dav-is replied:
How are we supposed to test the feature if we can't enable it?

styfle replied:
@dav-is There are two places `detectRoutes()` is called.

1. The `now dev` server which you commented on here
2. The `api-deployments` backend which will be modified once this PR is released

I am going to set the flag in `api-deployments` based on the now.json properties.

styfle committed:
Add export detectOutputDirectory

styfle committed:
Remove unused escape

styfle committed:
Disable feature by default

styfle committed:
Fix api test

styfle committed:
Add tests for featHandleMiss=true
